rhine-terminal/src/FRP/Rhine/Terminal.hs

```diff
@@ -0,0 +1,103 @@
+{- | Wrapper to write @terminal@ applications in Rhine, using concurrency.
+-}
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE RecordWildCards #-}
+module FRP.Rhine.Terminal
+  ( TerminalEventClock (..)
+  , flowTerminal
+  , terminalConcurrently
+  )
+  where
+
+-- base
+import Prelude hiding (putChar)
+import Control.Concurrent ()
+import Control.Concurrent.MVar ()
+import Data.IORef ()
+
+-- time
+import Data.Time.Clock ( getCurrentTime )
+
+-- terminal
+import System.Terminal
+    ( awaitEvent, runTerminalT, Event, Interrupt, TerminalT )
+import System.Terminal.Internal ( Terminal )
+
+-- transformers
+import Control.Monad.Trans.Reader
+
+-- rhine
+import FRP.Rhine.Clock.Proxy ()
+import FRP.Rhine
+import Control.Monad.Trans.Class (lift)
+
+-- * Terminal clock
+data TerminalEventClock = TerminalEventClock
+
+instance (Terminal t) => Clock (TerminalT t IO) TerminalEventClock
```

Comment:
```suggestion
instance (Terminal t, MonadIO m) => Clock (TerminalT t m) TerminalEventClock
```

Reply:
I'm running in circles trying to understand. Can we not do `(MonadInput m, MonadIO m) => Clock m TerminalEventClock`?

Reply:
Ah yes that makes a lot of sense. We only need `MonadInput` to `awaitEvent`.